### DISCUSSION TITLE
migrate wants to reset my DB when creating the initial migration

### DISCUSSION THREAD

**Discussion by blevine** [prisma/prisma#8860] [Q&A]
I'm clearly very confused.  I have an existing DB from which I created a Prisma schema using `prisma db pull`. I now want to create the initial migration and so I did: `prisma migrate dev --name initial-migration --create-only`.  However, I get back:

> 
> ? Drift detected: Your database schema is not in sync with your migration history.
> 
> The following is a summary of the differences between the expected database schema given your migrations files, and the actual schema of the database.
> 
> It should be understood as the set of changes to get from the expected schema to the actual schema.
> 
> [+] Added tables
> ...

which includes all tables in the DB followed by a long list of changes most of which are the addition of indexes and foreign keys.

Why am I seeing all these changes and why am I being warned about a DB reset when I added the --create-only option?

**Comment by ryands17** [comment]
Your development environment will be reset when you initialise migrations on an existing project as explained [here](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/add-prisma-migrate-to-a-project).

If you don't want to lose your development data, then you can do the following:

1. Create a new database which is empty and run `prisma migrate dev --name initial-migration`
2. After the migrations are created, baseline your current development database where you have your data as shown [here](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/add-prisma-migrate-to-a-project#baseline-your-production-environment).

This should help keeping your development data intact when performing a migration.

**Reply by blevine** [reply]
My apologies. Somehow I missed the entire "existing project" section in the docs.  Thank you for taking the time to explain.  I'd like to tell you about my development scenario which you might not have considered and that I admit might be an edge case.

I am integrating some custom user management software with the Keycloak identity management product. In order to do that, I will be adding tables to the Keycloak schema with references to existing Keycloak tables, but I will not be modifying Keycloak's own tables. If Prisma supported multiple schemas, I'd be putting my tables in their own schema.  For now, I'll just prefix the table names with something distinct.  My tables need to be in the same DB as the Keycloak tables to create efficient queries for my application. I may also need to access these tables from certain Keycloak plugins that I'll be writing.  So I _think_ (based on your recommendations) that what I want is to not apply the initial migration and so my development scenario looks something like this:

1. Install Keycloak (in this case in a Docker container). The installation process creates the Keycloak database with some initial data.
2. Create Prisma schema: db pull to create a Prisma schema that includes just the Keycloak tables. Note that I will _never_ modify these tables.
3. Initial migration: prisma migrate dev --name initial-migration --create-only
4. Baseline: prisma migrate resolve --applied <date-stamp>_initial-migration

From this point, I can add my own models to the Prisma schema and create/apply migrations to add the new tables to the DB.

Have I gotten this right?

**Reply by blevine** [reply]
Actually, I just tried this out and my steps are not quite correct. prisma migrate dev --name initial-migration --create-only **does** destroy the existing tables in the DB.  I guessed I missed your step 1 above about creating a new empty DB. I'll try applying the initial migration now to see if my DB schema is re-created exactly as it was before.

**Reply by ryands17** [reply]
The `--create-only` command will have to be done on a new clean DB and then you can baseline your existing development DB i.e. step 4.

**Reply by blevine** [reply]
OK, I get it.  This procedure is a bit cumbersome though and makes it difficult to bootstrap another developer. Although I suppose I could figure out how to script this. I guess what I'm really looking for is a way to create the migration table in an existing DB without blowing away the existing schema and without having to apply the initial migration.

**Reply by ryands17** [reply]
I wouldn't recommend using `db push` and then migrate as you would need to go over this process all over again. Using migrate right from the start and creating migrations for the changes you make is the way to go. Otherwise, you need to baseline your environment every time.

When you want to provide this setup to another developer, they just need to run `prisma migrate deploy` and their DB state will be the same as yours.

**Reply by blevine** [reply]
I see. I was initially concerned that using `prisma migrate deploy` would not create the shadow development DB. But I've re-read the docs and I see that the shadow DB is created/destroyed with each `prisma migrate dev` command.  So bootstrapping another developer with `prisma migrate deploy` seems workable.  I'll play with this a bit more.  Thanks for sticking with me while I figure this out! :)

**Reply by blevine** [reply]
Still running into some problems here. After generating the initial migration, I re-installed Keycloak to allow it to re-create it's database.  I then did `npx prisma migrate resolve `which added the initial migration to the database without executing it.  I then  did `npx prisma migrate dev`. I got "...drift detected..." with numerous diffs that look like:

> [*] Changed the `user_attribute` table
>   [-] Removed foreign key on columns (user_id)
>   [+] Added foreign key on columns (user_id)

so really no difference, but remove and adding back the same schema and still causing a reset.  I would have expected no differences and no reset.

Thoughts?

**Reply by blevine** [reply]
I may have worked around this although I'm not completely sure how.  I ended up doing a `pg_dump --schema-only` and putting the output in my initial migration file.  I deleted the existing row in the _prisma_migrations table and did `npx prisma migrate resolve`. After that, I did `npx prisma migrate dev` and didn't get any drift or -/+ changes as I described above. 

I was trying lots of different things to get this to work.  So tomorrow (because it's getting late here), I'm going to start from scratch and try these steps again to ensure I didn't miss anything. I'm still not sure what I did wrong that caused the schema drift and -/+ changes though. But for my use-case it might turn out that generating an initial schema with the output from pg_dump is the best way to go.

**Reply by blevine** [reply]
Ran into one more problem.  I added new table to schema.prisma:

```
model ab_camp_user {
  id      String      @id @default(uuid())
  user    user_entity @relation(fields: [user_id], references: [id])
  user_id String      @db.VarChar(36)
}
```

which generated the following migration:

```
-- CreateTable
CREATE TABLE "ab_camp_user" (
    "id" TEXT NOT NULL,
    "user_id" VARCHAR(36) NOT NULL,
    PRIMARY KEY ("id")
);

-- AddForeignKey
ALTER TABLE "ab_camp_user" ADD FOREIGN KEY ("user_id") REFERENCES "user_entity"("id") ON DELETE CASCADE ON UPDATE CASCADE;

-- CreateIndex
CREATE UNIQUE INDEX "ab_camp_user_user_id_unique" ON "ab_camp_user"("user_id");
```
all of which seems correct.

`prisma migrate dev `ran fine, but then asked me for a migration name which I didn't expect.  I entered a name and an additional migration was generated and executed:

```
-- DropIndex
DROP INDEX "ab_camp_user_user_id_unique";
```

I ran this scenario again and saw that after the original migration was applied, but before I entered a name for the 3rd migration, the ab_camp_user_user_id_unique index was correctly created.

Why does migrate think that it needs to remove that index?

**Reply by blevine** [reply]
Never mind.  I had defined the ab_camp_user field on the related model (user_entity) incorrectly.  The migration now works without prompting me for a name.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
